runtime/runtime-params-estimator/src/vm_estimator.rs

```diff
@@ -130,6 +135,128 @@ pub fn load_and_compile(
+    let start = start_count(gas_metric);
+    let vm_config = VMConfig::default();
+    let result = precompile_contract_vm(vm_kind, &contract, &vm_config, cache);
+    assert!(result.is_ok(), "Compilation failed");
+    let end = end_count(gas_metric, &start);
+    end
+}
+
+/// Returns `(a, b)` - approximation coefficients for formula `a + b * x`
+/// where `x` is is the contract size in bytes. Practically, we compute upper bound
+/// of this approximation, assuming that all contract consists of code only.
+fn precompilation_cost(gas_metric: GasMetric, vm_kind: VMKind) -> (Ratio<i128>, Ratio<i128>) {
+    let workdir = tempfile::Builder::new().prefix("runtime_testbed").tempdir().unwrap();
+    let store = create_store(&get_store_path(workdir.path()));
+    let cache_store = Arc::new(StoreCompiledContractCache { store });
+    let cache: Option<&dyn CompiledContractCache> = Some(cache_store.as_ref());
+
+    let mut xs = vec![];
+    let mut ys = vec![];
+
+    // We use core-contracts, e2f60b5b0930a9df2c413e1460e179c65c8876e3.
+    // File 341191, code 279965, data 56627.
+    let data = include_bytes!("../test-contract/res/lockup_contract.wasm");
+    let contract = ContractCode::new(data.to_vec(), None);
+    xs.push(data.len() as u64);
+    ys.push(measure_contract(vm_kind, gas_metric, &contract, cache));
+
+    // File 257516, code 203545, data 50419.
+    let data = include_bytes!("../test-contract/res/staking_pool.wasm");
+    let contract = ContractCode::new(data.to_vec(), None);
+    xs.push(data.len() as u64);
+    ys.push(measure_contract(vm_kind, gas_metric, &contract, cache));
+
+    // File 135358, code 113152, data 19520.
+    let data = include_bytes!("../test-contract/res/voting_contract.wasm");
+    let contract = ContractCode::new(data.to_vec(), None);
+    xs.push(data.len() as u64);
+    ys.push(measure_contract(vm_kind, gas_metric, &contract, cache));
+
+    // File 124250, code 103473, data 18176.
+    let data = include_bytes!("../test-contract/res/whitelist.wasm");
+    let contract = ContractCode::new(data.to_vec(), None);
+    xs.push(data.len() as u64);
+    ys.push(measure_contract(vm_kind, gas_metric, &contract, cache));
+
+    // Least squares method.
+    let n = xs.len();
+    let n128 = n as i128;
+
+    let mut sum_prod = 0 as i128; // Sum of x * y.
+    for i in 0..n {
+        sum_prod = sum_prod + (xs[i] as i128) * (ys[i] as i128);
+    }
+
+    let mut sum_x = 0 as i128; // Sum of x.
+    for i in 0..n {
+        sum_x = sum_x + (xs[i] as i128);
+    }
+
+    let mut sum_y = 0 as i128; // Sum of y.
+    for i in 0..n {
+        sum_y = sum_y + (ys[i] as i128);
+    }
+
+    let mut sum_x_square = 0 as i128; // Sum of x^2.
+    for i in 0..n {
+        sum_x_square = sum_x_square + (xs[i] as i128) * (xs[i] as i128);
+    }
+
+    let b = Ratio::new(n128 * sum_prod - sum_x * sum_y, n128 * sum_x_square - sum_x * sum_x);
+    let a = Ratio::new(sum_y * b.denom() - b.numer() * sum_x, n128 * b.denom());
```

Comment:
I use a differently stated formula in my version, and I pen&paper confiremed that they are equivalent.

Reply:
Yeah, this one denom is multiplied to both